src/scheduler/node_info.c

```diff
@@ -160,6 +160,63 @@
 /* name of the last node a job ran on - used in smp_dist = round robin */
 static char last_node_name[PBS_MAXSVRJOBID];
 
+void
+query_node_info_chunk(th_data_query_ninfo *data)
+{
+	struct batch_status *nodes;
+	struct batch_status *cur_node;
+	node_info **ninfo_arr;
+	server_info *sinfo;
+	node_info *ninfo;
+	int i;
+	int nidx;
+	int start;
+	int end;
+	int num_nodes_chunk;
+
+	nodes = data->nodes;
+	sinfo = data->sinfo;
+	start = data->sidx;
+	end = data->eidx;
+	num_nodes_chunk = end - start + 1;
+
+	if ((ninfo_arr = (node_info **) malloc((num_nodes_chunk + 1) * sizeof(node_info *))) == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		data->err = 1;
+		return;
+	}
+	ninfo_arr[0] = NULL;
+
+	/* Move to the linked list item corresponding to the 'start' index */
+	for (cur_node = nodes, i = 0; i < start && cur_node != NULL; cur_node = cur_node->next, i++)
+		;
+
+	for (i = start, nidx = 0; i <= end && cur_node != NULL; cur_node = cur_node->next, i++) {
+		/* get node info from the batch_status */
+		if ((ninfo = query_node_info(cur_node, sinfo)) == NULL) {
+			free_nodes(ninfo_arr);
+			data->err = 1;
+			return;
+		}
+
+		if (node_in_partition(ninfo, sinfo->partitions)) {
+			/* get node info from mom */
+			if (talk_with_mom(ninfo)) {
```

Comment:
talk_with_mom() might not be thread safe. Especially, it calls open_rm() which does not look like thread safe. Either we ensure that it is thread safe all through the call stack or we apply locks around it. However, serializing around an IO operation is generally a bad idea and could kill any performance benefit from the parallelization itself.

Reply:
thanks for catching this!